hypothesis-python/src/hypothesis/strategies/_internal/core.py

```diff
@@ -1577,9 +1577,29 @@ def as_strategy(strat_or_callable, thing, final=True):
             "Could not resolve %r to a strategy; consider "
             "using register_type_strategy" % (thing,)
         )
-    # Finally, try to build an instance by calling the type object
+    # Finally, try to build an instance by calling the type object.  Unlike builds(),
+    # this block *does* try to infer strategies for arguments with default values.
+    # That's because of the semantic different; builds() -> "call this with ..."
+    # so we only infer when *not* doing so would be an error; from_type() -> "give
+    # me arbitrary instances" so the greater variety is acceptable.
+    # And if it's *too* varied, express your opinions with register_type_strategy()
     if not isabstract(thing):
-        return builds(thing)
+        try:
+            hints = get_type_hints(thing)
+            params = signature(thing).parameters
+        except Exception:
+            return builds(thing)
+        kwargs = {}
+        for k, p in params.items():
+            if (
+                k in hints
+                and k not in ("self", "return")
```

Comment:
What about `cls`?

Reply:
One more thing: will this work correctly?

```python
def some(a: int, self: int = 0):
    return a + self
```

Reply:
Both good questions!  I don't expect `cls` to be bound *in `__init__`* the same way `self` often is where we can't really inspect it, so no special handling there.

And for the latter, not yet - but I'll add it as a test case, and only handle `self` specially when it's the first param name.

Reply:
Turns out that `inspect.signature` is smart enough to handle real `self` arguments for us, so we don't need any special handling at all!